### PR TITLE
Key Vault management Swagger spec

### PR DESCRIPTION
**Pull request by pomortaz**

### CODE REVIEW

**Comment by azurecla** [comment]
Hi __@pomortaz__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <p>
        It looks like you're working at Microsoft (pomortaz). If you're full-time, we DON'T require a contribution license agreement.
    </p>
    <p>
        If you are a vendor, DO please sign the electronic contribution license agreement. It will take 2 minutes and there's no faxing! https://cla.azure.com.
    </p>

TTYL, AZPRBOT;

**Comment by amarzavery** [comment]
This is awesome. We have the Key Vault Swagger spec Yay!!!

**Comment by pomortaz** [comment]
This is for management. The Key Vault data plane is still to come.

**Comment by amarzavery** [comment]
Ahh too early.. nonetheless something is better than nothing ;)

File: arm-keyvault/2015-06-01/swagger/keyvault.json [hunk 1]
```diff
@@ -0,0 +1,463 @@
...
+          },
+          {
+            "$ref": "#/parameters/SubscriptionIdParameter"
+          }
+        ],
+        "responses": {
+          "201": {
+            "description": "Created or updated vault",
+            "schema": {
+              "$ref": "#/definitions/Vault"
```
Comment (amarzavery) [inline arm-keyvault/2015-06-01/swagger/keyvault.json]: Is this a long running operation? If yes then please add the long running extension else ignore this comment.

File: arm-keyvault/2015-06-01/swagger/keyvault.json [hunk 2]
```diff
@@ -0,0 +1,463 @@
...
+            }
+          },
+          "200": {
+            "description": "Created or updated vault",
+            "schema": {
+              "$ref": "#/definitions/Vault"
+            }
+          }
+        }
```
Comment (amarzavery) [inline arm-keyvault/2015-06-01/swagger/keyvault.json]: default response is used for defining errors. Even if the default response is not specified, AutoRest generated code will deserialize the code and message properties of the error response. However, if the service is sending more information in the error response then it is recommended to define the default response for every operation. Otherwise, the extra information will not be deserialized.

File: arm-keyvault/2015-06-01/swagger/keyvault.json [hunk 3]
```diff
@@ -0,0 +1,463 @@
...
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/resources": {
+      "get": {
+        "tags": [
+          "Vaults"
+        ],
+        "operationId": "Vaults_List",
```
Comment (amarzavery) [inline arm-keyvault/2015-06-01/swagger/keyvault.json]: We are striving for a consistent experience across all the services in Azure to name some standard operations in the same way. The end user benefits from this. This should be named as **"Vaults_ListByResourceGroup"**. If there is a path that lists resources across subscriptions then that should be named as **Vaults_ListBySubscription"**

This issue https://github.com/Azure/azure-rest-api-specs/issues/244 will provide you more context.

File: arm-keyvault/2015-06-01/swagger/keyvault.json [hunk 4]
```diff
@@ -0,0 +1,463 @@
...
+          "nextLinkName": "nextLink"
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/resources": {
+      "get": {
+        "tags": [
+          "Vaults"
+        ],
+        "operationId": "Vaults_ListAll",
```
Comment (amarzavery) [inline arm-keyvault/2015-06-01/swagger/keyvault.json]: **"Vaults_ListBySubscription"**

File: arm-keyvault/2015-06-01/swagger/keyvault.json [hunk 5]
```diff
@@ -0,0 +1,463 @@
...
+            "type": "integer",
+            "format": "int32",
+            "description": "Maximum number of results to return."
+          },
+          {
+            "name": "api-version",
+            "in": "query",
+            "required": true,
+            "type": "string",
+            "enum": [ "2015-11-01" ],
```
Comment (amarzavery) [inline arm-keyvault/2015-06-01/swagger/keyvault.json]: This looks incorrect. Shouldn't api-version be `2015-06-01`, the one defined globally "version": "2015-06-01" . Is there a reason to define this locally and not use the globally defined api-version parameter?

File: arm-keyvault/2015-06-01/swagger/keyvault.json [hunk 6]
```diff
@@ -0,0 +1,463 @@
...
```
Comment (amarzavery) [inline arm-keyvault/2015-06-01/swagger/keyvault.json]: Same comment as above

File: arm-keyvault/2015-06-01/swagger/keyvault.json [hunk 7]
```diff
@@ -0,0 +1,463 @@
...
+            "description": "Azure Resource Manager Api Version."
+          },
+          {
+            "$ref": "#/parameters/SubscriptionIdParameter"
+          }
+        ],
+        "responses": {
+          "200": {
+            "description": "",
```
Comment (amarzavery) [inline arm-keyvault/2015-06-01/swagger/keyvault.json]: missing description

File: arm-keyvault/2015-06-01/swagger/keyvault.json [hunk 8]
```diff
@@ -0,0 +1,463 @@
...
```
Comment (amarzavery) [inline arm-keyvault/2015-06-01/swagger/keyvault.json]: missing description

File: arm-keyvault/2015-06-01/swagger/keyvault.json [hunk 9]
```diff
@@ -0,0 +1,463 @@
...
+        }
+      }
+    }
+  },
+  "definitions": {
+    "Sku": {
+      "properties": {
+        "family": {
+          "type": "string",
+          "description": "SKU family name"
```
Comment (amarzavery) [inline arm-keyvault/2015-06-01/swagger/keyvault.json]: Are there a set of possible/allowed values for Sku.family and Sku.name? If yes, then you should add the enum constraint with the set of allowed values. Please add the x-ms-enum extension and set modelAsString to true if more values will be added in the near future. false otherwise.

File: arm-keyvault/2015-06-01/swagger/keyvault.json [hunk 10]
```diff
@@ -0,0 +1,463 @@
...
+        "accessPolicies": {
+          "type": "array",
+          "items": {
+            "$ref": "#/definitions/AccessPolicyEntry"
+          },
+          "description": "Access policies for one or more principals"
+        },
+        "enabledForDeployment": {
+          "type": "boolean",
+          "description": "Enabled or disabled for deployment"
```
Comment (amarzavery) [inline arm-keyvault/2015-06-01/swagger/keyvault.json]: The following will give more info to the customer
> "Specifies whether Azure Virtual Machines are permitted to retrieve certificates stored as secrets from the key vault."
Got this from the [Key Vault MSDN REST API doc](https://msdn.microsoft.com/en-us/library/mt620025.aspx).
Please update the other descriptions as well.

File: arm-keyvault/2015-06-01/swagger/keyvault.json [hunk 11]
```diff
@@ -0,0 +1,463 @@
...
+        },
+        "tags": {
+          "type": "object",
+          "additionalProperties": {
+            "type": "string"
+          },
+          "description": "Resource tags"
+        },
+        "properties": {
+          "x-ms-client-flatten": false,
```
Comment (amarzavery) [inline arm-keyvault/2015-06-01/swagger/keyvault.json]: why is this set to false?

File: arm-keyvault/2015-06-01/swagger/keyvault.json [hunk 12]
```diff
@@ -0,0 +1,463 @@
...
+      "required": [
+        "location",
+        "properties"
+      ],
+      "x-ms-azure-resource": true
+    },
+    "Vault": {
+      "properties": {
+        "properties": {
+          "x-ms-client-flatten": false,
```
Comment (amarzavery) [inline arm-keyvault/2015-06-01/swagger/keyvault.json]: same as above

File: arm-keyvault/2015-06-01/swagger/keyvault.json [hunk 13]
```diff
@@ -0,0 +1,463 @@
...
+        "sku": {
+          "$ref": "#/definitions/Sku",
+          "description": "SKU details"
+        },
+        "accessPolicies": {
+          "type": "array",
+          "items": {
+            "$ref": "#/definitions/AccessPolicyEntry"
+          },
+          "description": "Access policies for one or more principals"
```
Comment (amarzavery) [inline arm-keyvault/2015-06-01/swagger/keyvault.json]: > An array of 0 to 16 identities that have access to the key vault. All identities in the array must use the same tenant ID as the key vault’s tenant ID.

The MSDN doc says that the user can provide upto 16 items in the accessPolicies array.
You may want to add "maxItems": 16 as a constraint.
Take a look at the possible constraints as per the swagger schema over [here](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md#patterned-fields-2). Once the page is loaded, scroll up a little and you should see the constraints.

File: arm-keyvault/2015-06-01/swagger/keyvault.json [hunk 14]
```diff
@@ -0,0 +1,463 @@
...
+  "definitions": {
+    "Sku": {
+      "properties": {
+        "family": {
+          "type": "string",
+          "description": "SKU family name"
+        },
+        "name": {
+          "type": "string",
+          "description": "SKU name"
```
Comment (amarzavery) [inline arm-keyvault/2015-06-01/swagger/keyvault.json]: https://msdn.microsoft.com/en-us/library/mt620025.aspx
> Specifies whether the key vault is a standard vault or a premium vault.

Looks like an enum candidate

File: arm-keyvault/2015-06-01/swagger/keyvault.json [hunk 15]
```diff
@@ -0,0 +1,463 @@
...
+          "200": {
+            "description": "Retrieved vault",
+            "schema": {
+              "$ref": "#/definitions/Vault"
+            }
+          }
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/resources": {
```
Comment (amarzavery) [inline arm-keyvault/2015-06-01/swagger/keyvault.json]: As per the [Key Vault MSDN REST API doc](https://msdn.microsoft.com/en-us/library/mt620027.aspx), the following path for listing all keyvaults in a resource group is supported

> https://management.azure.com/subscriptions/{subscription-id}/resourceGroups/{resource-group-name}/providers/Microsoft.KeyVault/vaults?api-version={api-version}

Then why do we need to use the generic resourcemanager path?

File: arm-keyvault/2015-06-01/swagger/keyvault.json (hunk 2 above)
Comment (pomortaz) [inline arm-keyvault/2015-06-01/swagger/keyvault.json]: For KV management, the error only contains code and message. So we use the built in exception.

Comment (amarzavery) [inline arm-keyvault/2015-06-01/swagger/keyvault.json]: perfect. sounds good.

File: arm-keyvault/2015-06-01/swagger/keyvault.json (hunk 3 above)
Comment (pomortaz) [inline arm-keyvault/2015-06-01/swagger/keyvault.json]: This would e a breaking change. We want to keep the name same as before.

File: arm-keyvault/2015-06-01/swagger/keyvault.json (hunk 5 above)
Comment (pomortaz) [inline arm-keyvault/2015-06-01/swagger/keyvault.json]: This version is for ARM and not Key Vault. We use ARM REST API directly to list vaults.
For listing vaults under a subscription, [this is how ARM is doing it](https://github.com/Azure/azure-rest-api-specs/blob/master/arm-resources/features/2015-12-01/swagger/features.json#L43) and we follow the same pattern.

File: arm-keyvault/2015-06-01/swagger/keyvault.json (hunk 3 above)
Comment (amarzavery) [inline arm-keyvault/2015-06-01/swagger/keyvault.json]: I understand that it is breaking change. IMHO, it would be ok to break them one time right now while moving from Hyak to AutoRest rather than breaking them again.

File: arm-keyvault/2015-06-01/swagger/keyvault.json (hunk 13 above)
Comment (pomortaz) [inline arm-keyvault/2015-06-01/swagger/keyvault.json]: Thanks. Good idea. Will add the constraint.

File: arm-keyvault/2015-06-01/swagger/keyvault.json (hunk 5 above)
Comment (amarzavery) [inline arm-keyvault/2015-06-01/swagger/keyvault.json]: Ah I see, while creating a resource in a generic way, one needs to provide the api-version of the RP. I assumed that the same logic would apply for listing resources.

File: arm-keyvault/2015-06-01/swagger/keyvault.json (hunk 11 above)
Comment (pomortaz) [inline arm-keyvault/2015-06-01/swagger/keyvault.json]: To avoid the unnecessary breaking change.

File: arm-keyvault/2015-06-01/swagger/keyvault.json (hunk 12 above)
Comment (pomortaz) [inline arm-keyvault/2015-06-01/swagger/keyvault.json]: To avoid the unnecessary breaking change.

File: arm-keyvault/2015-06-01/swagger/keyvault.json (hunk 10 above)
Comment (pomortaz) [inline arm-keyvault/2015-06-01/swagger/keyvault.json]: +1 thanks.

File: arm-keyvault/2015-06-01/swagger/keyvault.json (hunk 11 above)
Comment (amarzavery) [inline arm-keyvault/2015-06-01/swagger/keyvault.json]: If you do not want flattening then it is fine. You can remove it. AutoRest will not flatten it if the extension is not present. Setting it to false and not having the extension both mean the same :)

File: arm-keyvault/2015-06-01/swagger/keyvault.json (hunk 9 above)
Comment (pomortaz) [inline arm-keyvault/2015-06-01/swagger/keyvault.json]: Thanks. Will do. :)

File: arm-keyvault/2015-06-01/swagger/keyvault.json (hunk 4 above)
Comment (pomortaz) [inline arm-keyvault/2015-06-01/swagger/keyvault.json]: Will keep it like that to follow [ARM approach](https://github.com/Azure/azure-rest-api-specs/blob/master/arm-resources/features/2015-12-01/swagger/features.json#L43).

File: arm-keyvault/2015-06-01/swagger/keyvault.json (hunk 7 above)
Comment (pomortaz) [inline arm-keyvault/2015-06-01/swagger/keyvault.json]: Fixed.

File: arm-keyvault/2015-06-01/swagger/keyvault.json (hunk 15 above)
Comment (pomortaz) [inline arm-keyvault/2015-06-01/swagger/keyvault.json]: Thanks. Fixed. I missed it :+1:

File: arm-keyvault/2015-06-01/swagger/keyvault.json (hunk 1 above)
Comment (pomortaz) [inline arm-keyvault/2015-06-01/swagger/keyvault.json]: No, but thanks for pointing it out and for your thorough review. :)

**Comment by azurecla** [comment]
Hi __@pomortaz__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <p>
        It looks like you're working at Microsoft (pomortaz). If you're full-time, we DON'T require a contribution license agreement.
    </p>
    <p>
        If you are a vendor, DO please sign the electronic contribution license agreement. It will take 2 minutes and there's no faxing! https://cla.azure.com.
    </p>

TTYL, AZPRBOT;

**Comment by sphibbs** [comment]
@pomortaz We are moving all code to reflect the listby changes @amarzavery describes. ARM will be moving to. There are no good times for breaking changes, but there are logical times. 

Moving to swagger is a logical time to make breaking changes. 

We will be working with other partners to make these changes as well.

**Comment by pomortaz** [comment]
Thanks @sphibbs and @amarzavery for the explanation. If this is a recommended pattern, then we will adapt it.

**Comment by selvasingh** [comment]
@pomortaz - listBySubscription() should be re-labeled as list(). The recommended pattern is list() and listByResourceGroup()

**Comment by pomortaz** [comment]
@selvasingh thanks! I changed the ListBySubscription to List.

File: arm-keyvault/2015-06-01/swagger/keyvault.json [hunk 16]
```diff
@@ -0,0 +1,459 @@
...
+          "description": "The object ID of a user or service principal in the Azure Active Directory tenant for the vault. "
+        },
+        "applicationId": {
+          "type": "string",
+          "format": "uuid",
+          "description": " Application ID of the client making request on behalf of a principal"
+        },
+        "permissions": {
+          "$ref": "#/definitions/Permissions",
+          "description": "Permissions the identity has for keys and secrets"
```
Comment (tbombach) [inline arm-keyvault/2015-06-01/swagger/keyvault.json]: Per JSON Ref spec, any members besides $ref are supposed to be ignored: https://tools.ietf.org/html/draft-pbryan-zyp-json-ref-03#section-3

Could you remove the description field here and put it on the Permissions definition?

File: arm-keyvault/2015-06-01/swagger/keyvault.json [hunk 17]
```diff
@@ -0,0 +1,459 @@
...
+        "tags": {
+          "type": "object",
+          "additionalProperties": {
+            "type": "string"
+          },
+          "description": "The tags that will be assigned to the key vault. "
+        },
+        "properties": {
+          "$ref": "#/definitions/VaultProperties",
+          "description": "Properties of the vault"
```
Comment (tbombach) [inline arm-keyvault/2015-06-01/swagger/keyvault.json]: Same as above, could you move this description to the VaultProperties definition?

File: arm-keyvault/2015-06-01/swagger/keyvault.json [hunk 18]
```diff
@@ -0,0 +1,459 @@
...
+          "description": "The URI of the vault for performing operations on keys and secrets."
+        },
+        "tenantId": {
+          "type": "string",
+          "format": "uuid",
+          "description": "The Azure Active Directory tenant ID that should be used for authenticating requests to the key vault."
+        },
+        "sku": {
+          "$ref": "#/definitions/Sku",
+          "description": "SKU details"
```
Comment (tbombach) [inline arm-keyvault/2015-06-01/swagger/keyvault.json]: Same as above, could you move this description to the Sku definition?

File: arm-keyvault/2015-06-01/swagger/keyvault.json [hunk 19]
```diff
@@ -0,0 +1,459 @@
...
+        "location",
+        "properties"
+      ],
+      "x-ms-azure-resource": true
+    },
+    "Vault": {
+      "properties": {
+        "properties": {
+          "$ref": "#/definitions/VaultProperties",
+          "description": "Properties of the vault"
```
Comment (tbombach) [inline arm-keyvault/2015-06-01/swagger/keyvault.json]: Same as above, could you move this description to the VaultProperties definition?

File: arm-keyvault/2015-06-01/swagger/keyvault.json (hunk 16 above)
Comment (pomortaz) [inline arm-keyvault/2015-06-01/swagger/keyvault.json]: Thanks @tbombach. The description is not ignored in the generated code if used along with $ref. It is used as documentations for constructors and properties. If the description should be removed the autorest should be updated to include the referenced field documentation for properties and constructors. Until autorest is updated, I will keep the documentation as is.

File: arm-keyvault/2015-06-01/swagger/keyvault.json (hunk 19 above)
Comment (pomortaz) [inline arm-keyvault/2015-06-01/swagger/keyvault.json]: Answered above for theses comments.

**Comment by pomortaz** [comment]
@amarzavery and @selvasingh have you reached an agreement on the naming of List() vs ListBySubscription()? List() makes better sense as there is not argument passed to it. 
We are looking for a fixed recommendation that would be used across all management SDKs and would last long and not changing. :)

File: arm-keyvault/2015-06-01/swagger/keyvault.json (hunk 19 above)
Comment (selvasingh) [inline arm-keyvault/2015-06-01/swagger/keyvault.json]: Thank you @pomortaz

**Comment by amarzavery** [comment]
We will keep it as List(). Once the Java team acknowledges that the review is complete then the PR can be merged.

**Comment by anuchandy** [comment]
@amarzavery @selvasingh @pomortaz @jianghaolu 

Just had a discussion within the team and we agreed on 

[1] "List(`<no-param`)" and [2] "ListByResourceGroup(`string rgName`)"

File: arm-keyvault/2015-06-01/swagger/keyvault.json [hunk 20]
```diff
@@ -0,0 +1,459 @@
...
+            "name": "resourceGroupName",
+            "in": "path",
+            "required": true,
+            "type": "string",
+            "description": "The name of the Resource Group to which the vault belongs."
+          },
+          {
+            "name": "$top",
+            "in": "query",
+            "required": true,
```
Comment (jianghaolu) [inline arm-keyvault/2015-06-01/swagger/keyvault.json]: Why is this required?

File: arm-keyvault/2015-06-01/swagger/keyvault.json [hunk 21]
```diff
@@ -0,0 +1,459 @@
...
+            "type": "string",
+            "enum": [ "resourceType eq 'Microsoft.KeyVault/vaults'" ],
+            "description": "The filter to apply on the operation."
+          },
+          {
+            "name": "$top",
+            "in": "query",
+            "required": true,
```
Comment (jianghaolu) [inline arm-keyvault/2015-06-01/swagger/keyvault.json]: Same here.

Comment (pomortaz) [inline arm-keyvault/2015-06-01/swagger/keyvault.json]: Thanks @jianghaolu for catching that. The $top is not required. I fixed it. :+1:

**Comment by amarzavery** [comment]
Looks like the spec is good to be merged.

**Comment by lmazuel** [comment]
FYI https://github.com/Azure/azure-sdk-for-python/pull/679

**Comment by vishrutshah** [comment]
FYI: Ruby azure_mgmt_key_vault sdk - https://github.com/Azure/azure-sdk-ruby/pull/24
